astro/src/content/blog/typescript-return-type-annotation.md

```diff
@@ -0,0 +1,199 @@
+- Create a union of possible inferred types at different call sites
+- Struggle to unify the type across multiple code paths
+
+### 3. Spread Operator and Index Signatures
+
+The spread operator combined with computed property names creates additional complexity:
+
+```typescript
+return {
+  ...acc,                    // Type of acc must be unified
+  [dateKey]: {              // Computed property - dynamic key
+    utilizationLevel: ...,  // Type must be consistent across all branches
+  },
+};
+```
+
+TypeScript must ensure that:
+
+- The spread `...acc` is type-safe
+- The new property `[dateKey]` is compatible with the accumulated type
+- All possible return values from the reducer have the same type
+
+When `getUtilizationLevel` lacks an explicit return type, TypeScript may conservatively widen to `string` in some branches to ensure type safety.
+
+### 4. Effect.gen and Contextual Typing
+
+The `Effect.gen` generator function adds another layer of complexity. TypeScript must:
+
+- Infer types through generator yields
+- Maintain type information across async boundaries
+- Handle the Effect monad's type transformations
+
+This can interfere with type narrowing, especially for functions called deep within the generator.
+
+### 5. Union Type Propagation
+
+The error message showing a union of two object shapes suggests TypeScript created different type inferences for different code paths:
+
+```typescript
+// Path 1: TypeScript successfully narrows
+{ utilizationLevel: "low" | "medium" | "high"; ... }
+
+// Path 2: TypeScript widens to string (fallback)
+{ utilizationLevel: string; ... }
+```
+
+This happens when the compiler cannot prove that all code paths produce the same type, so it creates a union of possibilities.
+
+## Key Takeaways
+
+### When to Add Explicit Return Types
+
+1. **Functions returning union types of literals** - Always annotate these explicitly
+2. **Functions used in nested contexts** - Reducers, generators, higher-order functions
+3. **Functions used across module boundaries** - Especially when exported
+4. **When you see unexpected type widening** - `string` instead of `"a" | "b" | "c"`
+
+### Benefits of Explicit Return Types
+
+1. **Predictable type inference** - No surprises from compiler heuristics
+2. **Better error messages** - Errors appear at the function definition, not at call sites
+3. **Self-documenting code** - The return type is immediately visible
+4. **Compiler performance** - Less work for the type checker
+5. **Refactoring safety** - Changes to implementation won't accidentally change the public type
+
+### Best Practice
+
+```typescript
+// ❌ Avoid: Implicit return type in complex scenarios
+function getStatus(code: number) {
+  if (code === 200) return "success";
+  if (code === 404) return "not_found";
+  return "error";
+}
+
+// ✅ Prefer: Explicit return type
+function getStatus(code: number): "success" | "not_found" | "error" {
+  if (code === 200) return "success";
```

Comment:
Have you tried just adding `as const` to all of the returned strings? If this helps, I would honestly consider this to be the best practice, and to me, it's just a happy coincidence that it worked for some cases at all without `as const`

Reply:
Well, that should also work, but for my personal taste it has some disadvantages:

- It requires you to "annotate" all return branches
- If you forget to set one of the return values as `const` you will get again the string inferred without any type error until you find an incompatible case, way further from this function
- It is not as easy to understand. The return type annotation is in one single place, and you can very easily see it. If you rely on `as const` yo need to find all the return places yourself